### PR TITLE
187-4 Add a consistency check for the tezos-specific fork detector

### PR DESCRIPTION
**Pull request by ivanopagano**
Related to #42 and #187 

The currently available detector works under the assumption that the search for the fork level will be fast enough for the node to remain "stable".
That means that if the node "jumps" to a different fork while the detector is executing the search, we might get in serious trouble.
For Tezos is sounds reasonable that this might occur, even if only by a remote chance.
To handle the situation a more robust detector is defined to use with tezos specifically (since it involved data that is block-specific), which does a double-check before committing to the identified fork level.

**NOTE** This is part of a long-running streak of PRs to define the overall fork handling strategy.
Before merging it we need to merge the previous PRs #896 #898 and rebase on the new `master` branch, after which the `draft` tag could be removed.

### CODE REVIEW

**Review by piotrkosecki — APPROVED** [review]
Looks good!

File: conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/ConsistentForkDetector.scala [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+    attemptSearch(maxAttempts)
+  }
+
+  /** We want to make sure that during the search execution, which probably handles
+    * multiple I/O calls to compare indexer and node results, the node has kept giving
+    * consistent results.
+    * It might happen that network calls to both services were not fast enough, to the
+    * point that the fork on which the reference node was at the time of detection changed
+    * during the algorthm execution. This might be a reasonable expectation in a highly
```
Comment (vishakh) [inline conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/ConsistentForkDetector.scala:74]: Typo: Algorithm

File: conseil-lorre/src/test/scala/tech/cryptonomic/conseil/indexer/tezos/ConsistentForkDetectorSpec.scala [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+import org.scalatestplus.scalacheck.ScalaCheckDrivenPropertyChecks
+import org.scalacheck.{Arbitrary, Gen}
+import org.scalacheck.Arbitrary.arbitrary
+import org.scalacheck.ScalacheckShapeless._
+import monocle.macros.GenLens
+import tech.cryptonomic.conseil.indexer.tezos.ConsistentForkDetector.ConsistentDetectionRepeatingFailure
+
+/** Here we verify the properties of the [[ConsistentForkDetector]] implementation.
+  * This spec works essentially in the same way as
+  * the [[tech.cryptonomic.conseil.indexer.ForkDetectorSpec]]
```
Comment (vishakh) [inline conseil-lorre/src/test/scala/tech/cryptonomic/conseil/indexer/tezos/ConsistentForkDetectorSpec.scala:16]: The package name is incorrect.

**Review by vishakh — APPROVED** [review]

File: conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/ConsistentForkDetector.scala (hunk 1 above)
Comment (ivanopagano) [reply conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/ConsistentForkDetector.scala:74]: Fixed, and I additionally rephrased the "locally consistent" definition, trying to make it easier to understand.

File: conseil-lorre/src/test/scala/tech/cryptonomic/conseil/indexer/tezos/ConsistentForkDetectorSpec.scala (hunk 2 above)
Comment (ivanopagano) [reply conseil-lorre/src/test/scala/tech/cryptonomic/conseil/indexer/tezos/ConsistentForkDetectorSpec.scala:16]: fixed

**Pull request merged**